Surface_mesh/include/CGAL/Surface_mesh/IO/OFF.h

```diff
@@ -345,6 +345,48 @@ bool read_OFF(std::istream& is,
   return res;
 }
 
+template <typename Point>
+bool read_OFF(std::istream& is,
+              Surface_mesh<Point>& sm)
+{
+  return read_OFF(is, sm, parameters::all_default());
+}
+template <typename Point,
+          typename CGAL_BGL_NP_TEMPLATE_PARAMETERS>
+bool read_OFF(const char* fname,
+              Surface_mesh<Point>& sm,
+              const CGAL_BGL_NP_CLASS& np)
+{
+  std::ifstream in(fname);
+  return read_OFF(in, sm, np);
+}
+
+template <typename Point,
+          typename CGAL_BGL_NP_TEMPLATE_PARAMETERS>
+bool read_OFF(const char* fname,
+              Surface_mesh<Point>& sm)
+{
+  return read_OFF(fname, sm, parameters::all_default());
+}
+
+
+template <typename Point,
+          typename CGAL_BGL_NP_TEMPLATE_PARAMETERS>
+bool read_OFF(std::string& fname,
+              Surface_mesh<Point>& sm,
+              const CGAL_BGL_NP_CLASS& np)
+{
+  return read_OFF(fname.c_str(), sm, np);
+}
+
+template <typename Point,
+          typename CGAL_BGL_NP_TEMPLATE_PARAMETERS>
+bool read_OFF(std::string& fname,
+              Surface_mesh<Point>& sm)
+{
+  return read_OFF(fname, sm, parameters::all_default());
+}
+
```

Comment:
There must be something really wrong with that API, if one needs to define so many overloads!

Reply:
WELL, normally BGL would take care of all the extra and only the overload that's already here is necessary, but with the CGAL::IO:: namespace and this ADL problem we discussed, it seems that now, ALL OF THE OVERLOADS need to be defined, so yes, there is something really wrong.

Reply:
Is there a conclusion on the "really wrong" aspect?

Reply:
There is no conclusion. I am sorry if any of you two felt attacked by my comment. It was just a statement, without any proposed solution.

The issue seems to be a problem a [double-dispatch](https://en.wikipedia.org/wiki/Double_dispatch), not with dynamic polymorphism but with templates. The main key of that double-dispatch is the face-graph type. So maybe the facing function templates in `CGAL::IO::` could be generic for the first parameter:

```c++
template <typename Input, typename Face_graph>
bool read_OFF(Input&& input,
              Face_graph& graph);

template <typename Input, typename Point
bool read_OFF(Input&& input,
              Surface_mesh<Point>& graph);
              
// ...
```
and then, those function templates could dispatch again on the first parameter `Input input` calling functions in another namespace (`CGAL::IO::internal::`?)...

But I would recommend to refactor that way only for CGAL-5.4. That seems too late for CGAL-5.3. And anyway, the refactoring would be compatible with the API of CGAL-5.3.


... I am also surprised to see non-const references as parameter: `std::string& fname`. That seems to limit who users can call the function, and it should be fixed, in my opinion.

Reply:
No, I was just confused how moving namespaces across all functions could introduce such ADL issues that require these new overloads. But I haven't looked into it in detail (and so I was just wondering if the problem was already solved).

The `string` parameter should indeed be a `const&`.